src/txHashBuilder.c

```diff
@@ -1466,9 +1545,168 @@ static void txHashBuilder_assertCanLeaveNetworkId(tx_hash_builder_t* builder)
+
+	//Append CollRet, we don't allow DatumHash
+	cbor_append_legacy_txOutput(builder, addressBuffer, addressSize, amount, false);
+	//TODO: update this to new txout format
+}
+
+static void txHashBuilder_assertCanLeaveCollateralReturn(tx_hash_builder_t* builder)
+{
+	_TRACE("state = %d", builder->state);
+
+	switch (builder->state) {
+	case TX_HASH_BUILDER_IN_COLLATERAL_RETURN:
+		break;
+	case TX_HASH_BUILDER_IN_NETWORK_ID:
+	case TX_HASH_BUILDER_IN_REQUIRED_SIGNERS:
+	case TX_HASH_BUILDER_IN_COLLATERALS:
+	case TX_HASH_BUILDER_IN_SCRIPT_DATA_HASH:
+	case TX_HASH_BUILDER_IN_MINT:
+	case TX_HASH_BUILDER_IN_VALIDITY_INTERVAL_START:
+	case TX_HASH_BUILDER_IN_AUX_DATA:
+	case TX_HASH_BUILDER_IN_WITHDRAWALS:
+	case TX_HASH_BUILDER_IN_CERTIFICATES:
+	case TX_HASH_BUILDER_IN_TTL:
+	case TX_HASH_BUILDER_IN_FEE:
+		txHashBuilder_assertCanLeaveRequiredSigners(builder);
+		ASSERT(!builder->includeCollateralReturn);
+		break;
+
+	default:
+		ASSERT(false);
+	}
+}
+
+void txHashBuilder_addTotalCollateral(tx_hash_builder_t *builder, uint64_t txColl)
+{
+	_TRACE("state = %d", builder->state);
+
+	txHashBuilder_assertCanLeaveCollateralReturn(builder);
+	ASSERT(builder->includeTotalCollateral);
+
+
+	// add TotalCollateral item into the main tx body map
+	BUILDER_APPEND_CBOR(CBOR_TYPE_UNSIGNED, TX_BODY_KEY_TOTAL_COLLATERAL);
+	BUILDER_APPEND_CBOR(CBOR_TYPE_UNSIGNED, txColl);
+
+	builder->state = TX_HASH_BUILDER_IN_TOTAL_COLLATERAL;
+}
+
+static void txHashBuilder_assertCanLeaveTotalCollateral(tx_hash_builder_t* builder)
+{
+	_TRACE("state = %d", builder->state);
+
+	switch (builder->state) {
+	case TX_HASH_BUILDER_IN_TOTAL_COLLATERAL:
+		break;
+	case TX_HASH_BUILDER_IN_COLLATERAL_RETURN:
+	case TX_HASH_BUILDER_IN_NETWORK_ID:
+	case TX_HASH_BUILDER_IN_REQUIRED_SIGNERS:
+	case TX_HASH_BUILDER_IN_COLLATERALS:
+	case TX_HASH_BUILDER_IN_SCRIPT_DATA_HASH:
+	case TX_HASH_BUILDER_IN_MINT:
+	case TX_HASH_BUILDER_IN_VALIDITY_INTERVAL_START:
+	case TX_HASH_BUILDER_IN_AUX_DATA:
+	case TX_HASH_BUILDER_IN_WITHDRAWALS:
+	case TX_HASH_BUILDER_IN_CERTIFICATES:
+	case TX_HASH_BUILDER_IN_TTL:
+	case TX_HASH_BUILDER_IN_FEE:
+		txHashBuilder_assertCanLeaveRequiredSigners(builder);
+		ASSERT(!builder->includeTotalCollateral);
+		break;
+
+	default:
+		ASSERT(false);
+	}
+}
+
+void txHashBuilder_enterReferenceInputs(tx_hash_builder_t* builder )
+{
+	_TRACE("state = %d", builder->state);
+
+	txHashBuilder_assertCanLeaveTotalCollateral(builder);
+	// we don't allow an empty list for an optional item
+	ASSERT(builder->remainingReferenceInputs > 0);
+	{
+		// Enter reference inputs
+		BUILDER_APPEND_CBOR(CBOR_TYPE_UNSIGNED, TX_BODY_KEY_REFERENCE_INPUTS);
+		BUILDER_APPEND_CBOR(CBOR_TYPE_ARRAY, builder->remainingReferenceInputs);
+	}
+	builder->state = TX_HASH_BUILDER_IN_REFERENCE_INPUTS;
+
+}
+void txHashBuilder_addReferenceInputs(
+        tx_hash_builder_t* builder,
+        const uint8_t* utxoHashBuffer, size_t utxoHashSize,
```

Comment:
Since we use inputs in various places, it might make sense to create a structure instead of passing a triple of arguments all around. We actually already do have the type: `sign_tx_transaction_input_t`. It just needs to be moved to `cardano.h` and renamed to, say, `tx_input_t`.